python/ray/tests/test_multi_tenancy.py

```diff
@@ -240,30 +240,71 @@ def foo():
 def test_not_killing_workers_that_own_objects(shutdown_only):
     # Set the small interval for worker capping
     # so that we can easily trigger it.
-    ray.init(num_cpus=2, _system_config={"kill_idle_workers_interval_ms": 10})
+    ray.init(
+        num_cpus=1,
+        _system_config={
+            "kill_idle_workers_interval_ms": 10,
+            "worker_lease_timeout_milliseconds": 0
+        })
 
-    expected_num_workers = 8
+    expected_num_workers = 6
     # Create a nested tasks to start 8 workers each of which owns an object.
     @ray.remote
     def nested(i):
         # The task owns an object.
-        if i > expected_num_workers - 2:
+        if i >= expected_num_workers - 1:
             return [ray.put(np.ones(1 * 1024 * 1024, dtype=np.uint8))]
         else:
             return ([ray.put(np.ones(1 * 1024 * 1024, dtype=np.uint8))] +
                     ray.get(nested.remote(i + 1)))
 
     ref = ray.get(nested.remote(0))
-    num_workers = len(ray.state.workers())
+    num_workers = len(get_workers())
+
     # Wait for worker capping. worker capping should be triggered
     # every 10 ms, but we wait long enough to avoid a flaky test.
     time.sleep(1)
     ref2 = ray.get(nested.remote(0))
+
     # New workers shouldn't be registered because we reused the
     # previous workers that own objects.
-    assert num_workers == len(ray.state.workers())
-    assert len(ref) == expected_num_workers
+    assert num_workers == len(get_workers())
     assert len(ref2) == expected_num_workers
+    assert len(ref) == expected_num_workers
+
+
+def test_kill_idle_workers_that_are_behind_owned_workers(shutdown_only):
+    # When the first N idle workers own objects, and if we have N+N
+    # total idle workers, we should make sure other N workers are killed.
+    # It is because the idle workers are killed in the FIFO order.
+    N = 4
+    ray.init(
+        num_cpus=1,
+        _system_config={
+            "kill_idle_workers_interval_ms": 10,
+            "worker_lease_timeout_milliseconds": 0
+        })
+    expected_num_workers = N
+
+    @ray.remote
+    def nested(i):
+        if i >= (expected_num_workers * 2) - 1:
+            return [ray.put(np.ones(1 * 1024 * 1024, dtype=np.uint8))]
+        elif i >= expected_num_workers - 1:
+            return ([ray.put(np.ones(1 * 1024 * 1024, dtype=np.uint8))] +
+                    ray.get(nested.remote(i + 1)))
+        else:
+            return ([1] + ray.get(nested.remote(i + 1)))
+
+    # The first N workers don't own objects
+    # and the later N workers do.
+    ref = ray.get(nested.remote(0))
+    assert len(ref) == N * 2
+    num_workers = len(get_workers())
+    assert num_workers == N * 2
```

Comment:
Is this flaky? `num_workers == N + 1` is possible if `get_workers()` takes a long time.

Reply:
Changed the code properly.